daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/FNFunctions.scala

```diff
@@ -803,6 +805,84 @@ case class FNFloor(recipe: CompiledDPath, argType: NodeInfo.Kind) extends FNOneA
   }
 }
 
+case class DFDLXLeftShift(recipes: List[CompiledDPath], argType: NodeInfo.Kind) extends FNTwoArgs(recipes) {
+  override def computeValue(arg1: DataValuePrimitive, arg2: DataValuePrimitive, dstate: DState): DataValuePrimitive = {
+
+    val shiftLong = arg2.getLong
+    val shift = shiftLong.toInt
+    Assert.invariant(shift >= 0)
+    argType match {
+      case NodeInfo.Long => {
```

Comment:
Hmmm. Why dont we use ULong to represent unsignedLong?  Do we need it to be a subtype of java.lang.Number?

Reply:
Honestly, no idea. That's possible, but it's also possible it just a remnant of times before we added the Unsigned number library.

Should we create a bug to change our infoset to use ULong/UInt/UShort/UByte? Might be a bit of work, but would definitely make the code more obvious. And might also make ULong's more efficient since you don'thave the overhead of BigIntegers.

Reply:
I'll add a JIRA ticket for improvement. Our numbers are all boxed, and ULong and friends are defined to be non-boxing by extending AnyVal, not AnyRef. That's ok, and auto-boxed ULong is still cheaper than a bigInteger, which I believe uses a decimal representation.

Reply:
https://issues.apache.org/jira/browse/DAFFODIL-2511

Reply:
@mbeckerle What changes are expected from this PR?
Should I start with DAFFODIL-2511 and abandon these changes , or do I have to complete this for all the operators?

Reply:
I would not reccommend working on DAFFODIL-2511. That's an unrelated issue to this, and is probably going to be non-trivial to track down all the places where we should be using ULong/UInt/etc. I would just focus finishing up what you've got, keeping in mind that unsigned types in DPath are represented by the next larger signed type.

Reply:
PrimNumeric is already defined in [NodeInfo.scala](https://github.com/apache/daffodil/blob/master/daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/NodeInfo.scala#L475-L483) as a trait, so it sort acts much like an abstract class. Adding ``val width: MabyeInt`` to that existing trait, and then overriding it in the objects that mixin PrimNumeric (e.g. ``override val width = 32`` for NodeInfo.Int) should get the right behavior.

I think I would still use the same case/match you have where you match on argType with the NodeInfo.Foo cases to do the shifting. You can still use argType as a NodeInfo just as you did before. But since the argType must be a numeric primitive, it means you can **also** cast it to a PrimNumeric (which they all mixin) and get the width val.

I'm not sure why DState doesn't mixin ThrowsSDE, which would give it the schemaDefinitionUnless function. But there's probably a good reason. You can instead just do something like this:

```scala
if (shift >= primWidth) dstate.SDE("Left shift not supported for shift value greater than %s", argType.width)
```

Reply:
@stevedlawrence Thank You.
I have pushed the changes according to your suggestion.
Please have a look.